README.pod
=encoding utf-8

=for :stopwords Randy Stauner ACKNOWLEDGEMENTS repo repo's todo cpan testmatrix url
annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata
placeholders metacpan

=head1 NAME

Git::DescribeVersion - Use git-describe to show a repo's version

=head1 VERSION

version 1.015

=head1 SYNOPSIS

  use Git::DescribeVersion ();
  print Git::DescribeVersion->new({opt => 'value'})->version();

=head1 DESCRIPTION

Use C<git describe> to determine a git repo's version.

This is the main module,
though it's probably more useful run from the shell:

  $ git describe-version

The included C<git-describe-version> script
wraps L<Git::DescribeVersion::App>.

B<NOTE>: This module requires git version C<1.5.5> or greater.

The version is determined by counting the commits since the most recent tag
(matching the L</match_pattern>)
and using that count as the final part of the version.

So to create a typical three part version (C<v1.2.3>)
repo tags should be made of the first two parts (C<v1.2>)
and the number of commits counted by C<git-describe>
will become the third part (C<v1.2.35>).

=head1 METHODS

=head2 new

The constructor accepts a hash or hashref of options:

  Git::DescribeVersion->new({opt => 'value'});
  Git::DescribeVersion->new(opt1 => 'v1', opt2 => 'v2');

See L</OPTIONS> for an explanation of the available options.

=head2 format_version

Format the supplied version object
according to the L</format> attribute.

=head2 git

A method to wrap the git commands.
Attempts to use L<Git::Repository> or L<Git::Wrapper>.
Falls back to using backticks.

=head2 parse_version

A method to take the version parts found and return the end result.

Uses the L<version|version> module to parse.

=head2 version

The C<version> method is the main method of the class.
It attempts to return the repository version.

It will first use L</version_from_describe>.

If that fails it will try to simulate
the functionality with L</version_from_count_objects>
and will start the count from the L</first_version> option.

=head2 version_from_describe

Use C<git-describe> to count the number of commits since the last
tag matching L</match_pattern>.

It effectively calls

  git describe --match "${match_pattern}" --tags --long

If no matching tags are found (or some other error occurs)
it will return undef.

=head2 version_from_count_objects

Use C<git-count-objects> to count the number of commit objects
in the repository.  It then appends this count to L</first_version>.

It effectively calls

  git count-objects -v

and sums up the counts for 'count' and 'in-pack'.

=for Pod::Coverage git_backticks git_repository git_wrapper prepare_warning

=head1 OPTIONS

These options can be passed to L</new>:

=head2 directory

Directory in which git should operate.  Defaults to ".".

=head2 first_version

If the repository has no tags at all, this version
is used as the first version for the distribution.

Then git objects will be counted
and appended to create a version like C<v0.1.5>.

If set to C<undef> then L</version> will return undef
if L</version_from_describe> cannot determine a value.

Defaults to C<< v0.1 >>.

=head2 format

Specify the output format for the version number.

I had trouble determining the most reasonable names
for the formats so a few variations are possible.
(Pick the one which makes the most sense to you.)

=over 4

=item *

I<dotted>, I<normal>, I<v-string> or I<v>

for values like C<< v1.2.3 >>.

=item *

I<no-v-string> (or I<no-v> or I<no_v>)

to discard the opening C<v> for values like C<< 1.2.3 >>.

=item *

I<decimal>

for values like C<< 1.002003 >>.

=back

Defaults to I<decimal> for compatibility.

=head2 version_regexp

Regular expression that matches a tag containing
a version.  It must capture the version into C<$1>.

Defaults to C<< ([0-9._]+) >>
which will simply capture the first dotted-decimal found.
This matches tags like C<v0.1>, C<rev-1.2>
and even C<release-2.0-skippy>.

=head2 match_pattern

A shell-glob-style pattern to match tags.
This is passed to C<git-describe> to help it
find the right tag from which to count commits.

Defaults to C<< v[0-9]* >>.

=head1 HISTORY / RATIONALE

This module started out as a line in a Makefile:

  VERSION = $(shell (cd $(srcdir); \
    git describe --match 'v[0-9].[0-9]' --tags --long | \
    grep -Eo 'v[0-9]+\.[0-9]+-[0-9]+' | tr - . | cut -c 2-))

As soon as I wanted it in another Makefile
(in another repository) I knew I had a problem.

Then when I started learning L<Dist::Zilla>
I found L<Dist::Zilla::Plugin::Git::NextVersion>
but missed the functionality I was used to with C<git-describe>.

I started by forking L<Dist::Zilla::Plugin::Git> on github,
but realized that if I wrote the logic into a L<Dist::Zilla> plugin
it wouldn't be available to my git repositories that weren't Perl distributions.

So I wanted to extract the functionality to a module,
make a separate L<Dist::Zilla::Role::VersionProvider> plugin,
and include a quick version that could be run with a minimal
command line statement (so that I could put I<that> in my Makefiles).

=head1 TODO

=over 4

=item *

Allow using git-log to count commits affecting a subdirectory

=item *

Allow for more complex regexps (multiple groups) if there is a need.

=item *

Options for raising errors versus swallowing them?

=item *

Consider a dynamic installation to test C<`git --version`>.

=back

=head1 SEE ALSO

=over 4

=item *

L<Git::DescribeVersion::App>

=item *

L<Dist::Zilla::Plugin::Git::DescribeVersion>

=item *

L<Git::Repository> or L<Git::Wrapper>

=item *

L<http://www.git-scm.com>

=item *

L<version>

=back

=head1 SUPPORT

=head2 Perldoc

You can find documentation for this module with the perldoc command.

  perldoc Git::DescribeVersion

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

Search CPAN

The default CPAN search engine, useful to view POD in HTML format.

L<http://search.cpan.org/dist/Git-DescribeVersion>

=item *

RT: CPAN's Bug Tracker

The RT ( Request Tracker ) website is the default bug/issue tracking system for CPAN.

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Git-DescribeVersion>

=item *

CPAN Ratings

The CPAN Ratings is a website that allows community ratings and reviews of Perl modules.

L<http://cpanratings.perl.org/d/Git-DescribeVersion>

=item *

CPAN Testers

The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

L<http://www.cpantesters.org/distro/G/Git-DescribeVersion>

=item *

CPAN Testers Matrix

The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

L<http://matrix.cpantesters.org/?dist=Git-DescribeVersion>

=item *

CPAN Testers Dependencies

The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

L<http://deps.cpantesters.org/?module=Git::DescribeVersion>

=back

=head2 Bugs / Feature Requests

Please report any bugs or feature requests by email to C<bug-git-describeversion at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Git-DescribeVersion>. You will be automatically notified of any
progress on the request by the system.

=head2 Source Code


L<https://github.com/rwstauner/Git-DescribeVersion>

  git clone https://github.com/rwstauner/Git-DescribeVersion.git

=head1 AUTHOR

Randy Stauner <rwstauner@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Randy Stauner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Git::DescribeVersion

{{$NEXT}}

1.015     2012-08-22T03:11:08Z

  [Documentation]
  - Attempt to clarify intended usage in pod
    (module adds 3rd part to a 2-part tag (v1.2 => v1.2.35))
    to resolve rt-71626.
    Thanks to Matthew Astley for helping straighten this out.

  [Installation]
  - Halt installation if git version is less than 1.5.5

1.014     2011-11-06T00:34:15Z

  [Test Fixes]
  - Attempt to make t/git-repo.t compatible with older git versions
    in another way.

1.013     2011-10-26T14:45:17Z

  [Test Fixes]
  - Attempt to make t/git-repo.t compatible with older git versions

1.012     2011-10-24T14:43:20Z

  [Test Fixes]
  - Configure fake git user name/email for fake/test repo

1.011     2011-10-23T01:29:09Z

  [Tests]
  - Display git command output upon test failure (for debugging reports)

1.010     2011-10-19T20:31:43Z

  - Fallback to another git wrapper if one fails.
    Closes rt-71622. Thanks to Matthew (mca1001) for the idea.
  - Localize $/ to work around Git::Repository bug rt-71621.
  - Fixes for Windows

1.003     2011-07-15T00:00:55Z

  - Add git-describe-version executable
  - Make Git::DescribeVersion::App::run look in @ARGV as well

  - Improvements to documentation and tests
  - Superficial dist/release changes

1.001011  2010-12-17T04:25:14Z

  - Improve handling of trailing zeroes ("0th commit since tag")
  - Improvements to documentation and tests
  - use Module::Build

1.001005  2010-12-10T19:29:26Z

  - Require 0.82 of version.pm (instead of 0.77)
  - Improvements to documentation and tests

1.000010  2010-12-07T21:07:07Z

  - Add option to specify desired version format
    (dotted-decimal, decimal, or dotted w/o the leading 'v')
  - Enable module to work with Git::Wrapper, Git::Repository, or `git`
  - Improvements to documentation and tests

0.007006  2010-11-17T16:22:16Z

  - Moved Dist::Zilla::Plugin::Git::DescribeVersion to a separate dist

0.006013  2010-11-17T03:55:00Z

  - Add example gitconfig usage

0.006012  2010-11-15T21:27:00Z

  - Allow Git::DescribeVersion::App::run() to be called as a class method
  - Change default 'version_regexp' to catch a wider variety of tag patterns
  - Improvements to documentation and tests

0.005011  2010-11-09T19:34:00Z

  - Catch 'describe' errors and show a warning
  - Improvements to documentation and tests

0.004008  2010-10-25T06:09:00Z

  - Initial Release to CPAN


t/lib/GitDVTest.pm
# vim: set ts=2 sts=2 sw=2 expandtab smarttab:
package GitDVTest;
use strict;
use Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(
  &description
  &expect_warning
  &expectation
  &test_expectations
  &mock_gw
  @versions
  @commits
  @counts
);
use Test::More 0.96;
use Test::MockObject::Extends 1.09;
use version 0.82;

my $test_warn_mod = 'Test::Output';
my $test_warn = eval "use $test_warn_mod; 1";
$test_warn = 0 if $@;

# use sprintf for ease/clarity; use %d to turn '' into 0
sub description { sprintf '%s-%d-%s', $_[0], $_[1], 'gdeadbeef' }

sub expect_warning ($$;$) {
  my $sub = pop @_;
  my ($regexp, $message) = @_;
  $message ||= 'warning expected:';

  if( $test_warn ){
    stderr_like($sub, $regexp, $message);
  }
  else {
    # skip the test, but still run the coderef
    SKIP: { skip "$test_warn_mod required for testing warnings", 1; }
    # mention that we're expecting a warning
    diag($message);
    &$sub;
  }
}

sub expectation ($$$) {
  my ($gv, $version, $count) = @_;
  my ($tag, $dec, $dot, $regexp) = @$version;
  $tag = '~' if !defined $tag;
  $count ||= 0;

  # hack
  $gv->{version_regexp} = $regexp ||
    $Git::DescribeVersion::Defaults{version_regexp};

  my $exp    = defined $dec ? sprintf(($dec =~ /\./ ? "%s%03d" : "%s.%03d000"), $dec, $count): undef;
  my $dotted = defined $dot ? version->parse("$dot.$count")->normal : undef;
  my %values = (
    decimal => $exp,
    dotted  => $dotted,
    no_v    => (defined $dotted ? substr($dotted, 1) : undef),
  );
  return { map {
    ($_ => [$values{$_}, sprintf("describe %-15s in %s as %-15s",
      "$tag-$count", $_, defined $values{$_} ? $values{$_} : '~')])
  } keys %values };
}

sub test_expectations ($$$&) {
  my ($gv, $version, $commits, $sub) = @_;
  my $exps = expectation($gv, $version, $commits);
  while( my ($format, $expdesc) = each %$exps ){
    $gv->{format} = $format;
    $sub->(@$expdesc);
  }
}

sub mock_gw () {
  my $mock = Test::MockObject->new();
  $mock->fake_module('Git::Wrapper' => __version => sub { '--version' });
  $mock->mock(__version => sub { '--version' });
  $mock;
}

# Should we be using version->parse->numify
# instead of specifying the expectation explicitly?

# make sub-arrays like (['v0.1', '0.001', 'v0.1', 'version_regexp'])
our @versions = map { [(split(/\s+/))[1, 2, 3, 4]] } split(/\n/, <<TAGS);
  v0.1                      0.001     v0.1
  v0.001                    0.001     v0.1
  v1.2                      1.002     v1.2
  v1.20                     1.020     v1.20
  v1.200                    1.200     v1.200
  v1.02                     1.002     v1.2
  v1.002                    1.002     v1.2
  v1.2.3                    1.002003  v1.2.3
  v1.02.03                  1.002003  v1.2.3
  v1.002003                 1.2003    v1.2003
  v2.1                      2.001     v2.1
  v2.1234                   2.1234    v2.1234
  ver-0.012                 0.012     v0.12      ver-(.+)
  ver-0.012                 0.012     v0.12
  ver|3.222                 3.222     v3.222     ver\\|(.+)
  ver|3.222                 3.222     v3.222
  4.1-rel1021               4.001     v4.1       ([0-9.]+)-rel.+
  4.1-rel1021               4.001     v4.1
  4.1-rel10.21              10.021    v10.21     rel(\\S+)
  release-1.2-narf          1.002     v1.2
  release-1.4.2-narf        1.004     v1.4       \\w+-([0-9.]+)\\.\\d-narf
  release-1.2-narf          1         v1         \\w+-(\\d+)\\.\\d-narf
  SILLY1.4TAG               1.004     v1.4
  date-12.05-ver-10.21-foo  10.021    v10.21     date-[0-9.]+-ver-([0-9.]+)-\\w+
  date-12.05-ver-10.21-foo  12.005    v12.5
TAGS

our @commits = qw(0 8 12 49 99 135 999 1234);

# make sub-arrays like (['0', 'count: 0', 'size: 0'])
our @counts = map { [split(/\n/)] } split(/\n\n/, <<COUNTS);
204
count: 204
size: 816
in-pack: 0
packs: 0
size-pack: 0
prune-packable: 0
garbage: 0

1006
count: 604
size: 816
in-pack: 402

999
count: 999
in-pack: 0

322
count: 222
in-pack: 100

24
count: 24

7
in-pack: 7
COUNTS

1;


inc/GitVersion.pm
package # no_index
  inc::GitVersion;

# 1.5 renamed git-repo-config to git-config
# 1.5.5 added --match to git-describe
our @MIN = qw( 1 5 5 );

sub import {
  shift->require_minimum;
}

sub check_minimum {
  my ($class, $version) = @_;
  $version ||= $class->version;

  my @parts = split(/\./, ($version =~ m/(?:git\sversion\s)?(\d+(\.\d+)*)/)[0]);

  foreach my $i ( 0 .. $#MIN ){
    # if not equal
    if( my $cmp = $parts[$i] <=> $MIN[$i] ){
      # return true if version is sufficient (greater) false if not (lesser)
      return $cmp > 0;
    }
  }

  return 1; # not less or greater, must be equal
}

sub require_minimum {
  my ($class, $version) = @_;
  $version ||= $class->version;
  $class->check_minimum($version)
    or die <<MSG;
 # Git version 1.5.5+ required.
 # Found: $version
 #
 # If you believe this to be an error please submit a bug report
 # including the output of `git --version`.
MSG
}

sub version {
  chomp(my $v = qx/git --version/);
  $v;
}

1;


lib/Git/DescribeVersion/App.pm
# vim: set ts=2 sts=2 sw=2 expandtab smarttab:
use strict;
use warnings;

package Git::DescribeVersion::App;
# ABSTRACT: Run Git::DescribeVersion as one-line script

use Git::DescribeVersion ();
use Getopt::Long qw(GetOptions); # core

# for simplicity
our %Defaults = %Git::DescribeVersion::Defaults;

my %get_opt_spec = map {
  my $dash = $_;
  # accept --opt_name or --opt-name (but store it in the hash as opt_name)
  # don't duplicate single words ("format=s" rather than "format|format=s")
  ( $dash =~ tr/_/-/ ? "$_|$dash=s" : "$_=s" )
} keys %Defaults;

# simple: enable `perl -MGit::DescribeVersion::App -e run`
sub import {
  no strict 'refs'; ## no critic
  *{caller(0) . '::run'} = \&run;
}

sub options {
  # allow usage as Git::DescribeVersion::App->run()
  # (for consistency with other App's)
  # and simply discard the unused argument
  shift(@_) if @_ && $_[0] eq __PACKAGE__;

  my %env;
  foreach my $opt ( keys %Defaults ){
    # look for $ENV{GIT_DV_OPTION}
    my $eopt = "\UGIT_DV_$opt";
    $env{$opt} = $ENV{$eopt} if exists($ENV{$eopt});
  }

  my %argv;
  GetOptions(\%argv, 'help' => \&usage, %get_opt_spec)
    or usage();

  my %args = ref($_[0]) ? %{$_[0]} : @_;

  # order of importance: %ENV, @ARGV, @_
  return {%env, %argv, %args};
}

sub run {
  print Git::DescribeVersion->new(options(@_))->version, "\n";
}

sub usage {
  no warnings 'uninitialized';
  # show package name if script name is "-" or "-e"
  printf("%s %s\n\nOptions (and their default values):\n\n",
    ($0 =~ /^-e?$/ ? __PACKAGE__ : $0), __PACKAGE__->VERSION
  );

  foreach my $opt ( sort keys %Defaults ){
    (my $arg = $opt) =~ tr/_/-/;
    printf(qq[  --%-18s "%s"\n], $arg, $Defaults{$opt});
  }

  print "\nFor more information try `perldoc Git::DescribeVersion::App`\n";
  exit;
}

1;

=for Pod::Coverage options usage

=for test_synopsis
1;
__END__

=head1 SYNOPSIS

Print out the version from L<Git::DescribeVersion> in one line:

  perl -MGit::DescribeVersion::App -e run

Options can be passed as program arguments (C<@ARGV>):

  perl -MGit::DescribeVersion::App -e run --match-pattern "rev-*"

The C<@ARGV> form allows arguments
to be spelled with dashes instead of underscores.

Options can also be passed in a hash or hashref
just like L<Git::DescribeVersion/new>:

  perl -MGit::DescribeVersion::App -e 'run(match_pattern => "rev-*")'

Or can be environment variables spelled like I<GIT_DV_OPTION>:

  export GIT_DV_MATCH_PATTERN="rev-*"
  perl -MGit::DescribeVersion::App -e run

This (hopefully) makes it easy for you to write
the alias, function, Makefile or script that does exactly what you want.

If not, feel free to send me suggestions (or patches)
that you think would make it simpler or more powerful.

=method run

Convenience method for writing one-liners.

Exported to main package.

Accepts arguments in a hash or hashref
which are passed to the constructor.

Also looks for arguments in %ENV and @ARGV.

See L</SYNOPSIS>.

=head1 SEE ALSO

=for :list
* L<Git::DescribeVersion>

=cut


lib/Git/DescribeVersion.pm
# vim: set ts=2 sts=2 sw=2 expandtab smarttab:
use strict;
use warnings;

package Git::DescribeVersion;
# ABSTRACT: Use git-describe to show a repo's version

use Carp (); # core
use version 0.82 ();

our %Defaults = (
  first_version   => 'v0.1',
  match_pattern   => 'v[0-9]*',
  format      => 'decimal',
  version_regexp  => '([0-9._]+)'
);

# Git::Repository is easier to install than Git::Wrapper
my @delegators = qw(
  git_repository
  git_wrapper
  git_backticks
);

=method new

The constructor accepts a hash or hashref of options:

  Git::DescribeVersion->new({opt => 'value'});
  Git::DescribeVersion->new(opt1 => 'v1', opt2 => 'v2');

See L</OPTIONS> for an explanation of the available options.

=cut

sub new {
  my $class = shift;
  # accept a hash or hashref
  my %opts = ref($_[0]) ? %{$_[0]} : @_;
  my $self = {
    %Defaults,
    # restrict accepted arguments
    map { $_ => $opts{$_} } grep { exists($opts{$_}) } keys %Defaults
  };

  $self->{directory} = $opts{directory} || '.';
  bless $self, $class;

  # accept a Git::Repository or Git::Wrapper object (or command to exec)
  # or a simple '1' (true value) to indicate which one is desired
  foreach my $mod ( @delegators ){
    if( $opts{$mod} ){
      $self->{git} = $mod;
      # if it's just a true value leave it blank so we create later
      # TODO: should this be checking ref?
      $self->{$mod} = $opts{$mod}
        unless $opts{$mod} eq '1';
      # test that requested method "works"
      eval { $self->$mod('--version') };
      if( $@ ){
        Carp::carp qq[Failed to execute $mod (will attempt other methods): $@];
        delete @$self{(git => $mod)};
      }
    }
  }
  return $self;
}

=method format_version

Format the supplied version object
according to the L</format> attribute.

=cut

sub format_version {
  my ($self, $vobject) = @_;
  my $format = $self->{format} =~ /dot|normal|v|string/ ? 'normal' : 'numify';
  my $version = $vobject->$format;
  $version =~ s/^v// if $self->{format} =~ /no.?v/;
  return $version;
}

=method git

A method to wrap the git commands.
Attempts to use L<Git::Repository> or L<Git::Wrapper>.
Falls back to using backticks.

=cut

# NOTE: the git* subs are called in list context

sub git {
  my ($self) = @_;
  unless( $self->{git} ){
    for my $method ( @delegators ){
      $self->{git} ||= eval {
        # confirm method works (without dying)
        $self->$method('--version');
        $method;
      };
    }
    Carp::croak("All git methods failed.  Is `git` installed?\n".
      "Consider installing Git::Repository or Git::Wrapper.\n")
      unless $self->{git};
  }
  goto &{$self->{git}};
}

sub git_backticks {
  my ($self, $command, @args) = @_;
  warn("'directory' attribute not supported when using backticks.\n" .
    "Consider installing Git::Repository or Git::Wrapper.\n")
      if $self->{directory} && $self->{directory} ne '.';

  @args = map { ref $_ ? @$_ : $_ } @args;

  @args = map { quotemeta } @args
    unless $^O eq 'MSWin32';

  my $exec = join(' ',
      # the external app to run
      ($self->{git_backticks} ||= 'git'),
      $command,
      @args
  );

  return (`$exec`);
}

sub git_repository {
  my ($self, $command, @args) = @_;
  # Git::Repository 1.22 fails with alternate $/ (rt-71621)
  local $/ = "\n";
  (
    $self->{git_repository} ||=
    do {
      require Git::Repository;
      Git::Repository->new(work_tree => $self->{directory})
    }
  )
    ->run($command,
      map { ref $_ ? @$_ : $_ } @args
    );
}

sub git_wrapper {
  my ($self, $command, @args) = @_;
  $command =~ tr/-/_/;
  (
    $self->{git_wrapper} ||=
    do {
      require Git::Wrapper;
      Git::Wrapper->new($self->{directory})
    }
  )
    ->$command({
      map { ($$_[0] =~ /^-{0,2}(.+)$/, $$_[1]) }
        map { ref $_ ? $_ : [$_ => 1] } @args
    });
}


=method parse_version

A method to take the version parts found and return the end result.

Uses the L<version|version> module to parse.

=cut

sub parse_version {
  my ($self, $prefix, $count) = @_;

  # This is unlikely as it should mean that both git commands
  # returned unexpected output.  If it does happen, don't die
  # trying to parse it, default to first_version.
  $prefix = $self->{first_version}
    unless defined $prefix;
  $count ||= 0;

  # If still undef (first_version explicitly set to undef)
  # don't die trying to parse it, just return nothing.
  unless( defined $prefix ){
    warn("Version could not be determined.\n");
    return;
  }

  # s//$1/ requires the regexp to be anchored.
  # Doing a match and then assigning to $1 does not.
  if( $self->{version_regexp} && $prefix =~ /$self->{version_regexp}/ ){
    $prefix = $1;
  }

  my $vstring = "v$prefix.$count";

  # quote 'version' to reference the module and not call the local sub
  my $vobject = eval {
    # don't even try to parse it if it doesn't look like a version
    'version'->parse($vstring)
      if version::is_lax($vstring);
  };

  # Don't die if it's not parseable, just return nothing.
  if( my $error = $@ || !$vobject ){
    $error = $self->prepare_warning($error);
    warn("'$vstring' is not a valid version string.\n$error");
    return;
  }

  return $self->format_version($vobject);
}

# normalize error message

sub prepare_warning {
  my ($self, $error) = @_;
  return '' unless $error;
  $error =~ s/ at \S+?\.pm line \d+\.?\s*$//;
  chomp($error);
  return $error . "\n";
}

=method version

The C<version> method is the main method of the class.
It attempts to return the repository version.

It will first use L</version_from_describe>.

If that fails it will try to simulate
the functionality with L</version_from_count_objects>
and will start the count from the L</first_version> option.

=cut

sub version {
  my ($self) = @_;
  return $self->version_from_describe() ||
    $self->version_from_count_objects();
}

=method version_from_describe

Use C<git-describe> to count the number of commits since the last
tag matching L</match_pattern>.

It effectively calls

  git describe --match "${match_pattern}" --tags --long

If no matching tags are found (or some other error occurs)
it will return undef.

=cut

sub version_from_describe {
  my ($self) = @_;
  my ($ver) = eval {
    $self->git('describe',
      ['--match' => $self->{match_pattern}], qw(--tags --long)
    );
  };
  # usually you'll expect a tag to be found, so warn if it isn't
  if( my $error = $@ ){
    $error = $self->prepare_warning($error);
    warn("git-describe: $error");
  }

  # return nothing so we know to move on to count-objects
  return unless $ver;

  # ignore the -gSHA
  my ($tag, $count) = ($ver =~ /^(.+?)-(\d+)-(g[0-9a-f]+)$/);

  return $self->parse_version($tag, $count);
}

=method version_from_count_objects

Use C<git-count-objects> to count the number of commit objects
in the repository.  It then appends this count to L</first_version>.

It effectively calls

  git count-objects -v

and sums up the counts for 'count' and 'in-pack'.

=cut

sub version_from_count_objects {
  my ($self) = @_;
  my @counts = $self->git(qw(count-objects -v));
  my $count = 0;
  local $_;
  foreach (@counts){
    /(count|in-pack): (\d+)/ and $count += $2;
  }
  return $self->parse_version($self->{first_version}, $count);
}

1;

=for Pod::Coverage git_backticks git_repository git_wrapper prepare_warning

=for stopwords repo's todo

=head1 SYNOPSIS

  use Git::DescribeVersion ();
  print Git::DescribeVersion->new({opt => 'value'})->version();

=head1 DESCRIPTION

Use C<git describe> to determine a git repo's version.

This is the main module,
though it's probably more useful run from the shell:

  $ git describe-version

The included C<git-describe-version> script
wraps L<Git::DescribeVersion::App>.

B<NOTE>: This module requires git version C<1.5.5> or greater.

The version is determined by counting the commits since the most recent tag
(matching the L</match_pattern>)
and using that count as the final part of the version.

So to create a typical three part version (C<v1.2.3>)
repo tags should be made of the first two parts (C<v1.2>)
and the number of commits counted by C<git-describe>
will become the third part (C<v1.2.35>).

=head1 OPTIONS

These options can be passed to L</new>:

=head2 directory

Directory in which git should operate.  Defaults to ".".

=head2 first_version

If the repository has no tags at all, this version
is used as the first version for the distribution.

Then git objects will be counted
and appended to create a version like C<v0.1.5>.

If set to C<undef> then L</version> will return undef
if L</version_from_describe> cannot determine a value.

Defaults to C<< v0.1 >>.

=head2 format

Specify the output format for the version number.

I had trouble determining the most reasonable names
for the formats so a few variations are possible.
(Pick the one which makes the most sense to you.)

=for :list
* I<dotted>, I<normal>, I<v-string> or I<v>
for values like C<< v1.2.3 >>.
* I<no-v-string> (or I<no-v> or I<no_v>)
to discard the opening C<v> for values like C<< 1.2.3 >>.
* I<decimal>
for values like C<< 1.002003 >>.

Defaults to I<decimal> for compatibility.

=head2 version_regexp

Regular expression that matches a tag containing
a version.  It must capture the version into C<$1>.

Defaults to C<< ([0-9._]+) >>
which will simply capture the first dotted-decimal found.
This matches tags like C<v0.1>, C<rev-1.2>
and even C<release-2.0-skippy>.

=head2 match_pattern

A shell-glob-style pattern to match tags.
This is passed to C<git-describe> to help it
find the right tag from which to count commits.

Defaults to C<< v[0-9]* >>.

=head1 HISTORY / RATIONALE

This module started out as a line in a Makefile:

  VERSION = $(shell (cd $(srcdir); \
    git describe --match 'v[0-9].[0-9]' --tags --long | \
    grep -Eo 'v[0-9]+\.[0-9]+-[0-9]+' | tr - . | cut -c 2-))

As soon as I wanted it in another Makefile
(in another repository) I knew I had a problem.

Then when I started learning L<Dist::Zilla>
I found L<Dist::Zilla::Plugin::Git::NextVersion>
but missed the functionality I was used to with C<git-describe>.

I started by forking L<Dist::Zilla::Plugin::Git> on github,
but realized that if I wrote the logic into a L<Dist::Zilla> plugin
it wouldn't be available to my git repositories that weren't Perl distributions.

So I wanted to extract the functionality to a module,
make a separate L<Dist::Zilla::Role::VersionProvider> plugin,
and include a quick version that could be run with a minimal
command line statement (so that I could put I<that> in my Makefiles).

=head1 TODO

=for :list
* Allow using git-log to count commits affecting a subdirectory
* Allow for more complex regexps (multiple groups) if there is a need.
* Options for raising errors versus swallowing them?
* Consider a dynamic installation to test C<`git --version`>.

=head1 SEE ALSO

=for :list
* L<Git::DescribeVersion::App>
* L<Dist::Zilla::Plugin::Git::DescribeVersion>
* L<Git::Repository> or L<Git::Wrapper>
* L<http://www.git-scm.com>
* L<version>

=cut
